### PR TITLE
Use std::set instead of std::vector to indicate boundary attributes for BCs

### PR DESCRIPTION
**Pull request by jamiebramwell**
This changes the user-facing interface for solvers to use `std::set`s instead of `std::vector`s of indicator values to denote the boundary attributes for applying boundary conditions

### CODE REVIEW

**Comment by jamiebramwell** [comment]
LGTM

File: tests/serac_component_bc.cpp [hunk 1]
```diff
@@ -135,13 +132,9 @@ TEST(component_bc, qs_attribute_solve)
...
   NonlinearSolidSolver solid_solver(2, pmesh);
 
-  // define a boundary attribute array and initialize to 0
-  std::vector<int> ess_x_bdr(pmesh->bdr_attributes.Max(), 0);
-  std::vector<int> ess_y_bdr(pmesh->bdr_attributes.Max(), 0);
-
   // boundary attribute 1 (index 0) is fixed (Dirichlet) in the x direction
-  ess_x_bdr[0] = 1;
-  ess_y_bdr[1] = 1;
+  std::set<int> ess_x_bdr = {1};
```
Comment (white238) [inline tests/serac_component_bc.cpp:136]: School me please, what does 1 vs 2 represent/mean?

**Review by white238 — APPROVED** [review]

File: tests/serac_component_bc.cpp (hunk 1 above)
Comment (jamiebramwell) [reply tests/serac_component_bc.cpp:136]: The input mesh has different markers (1,2,etc) for each part of the boundary. This is saying the boundary with the '1' markers are of the essential x boundary condition type. You could have multiple markers for a single boundary, so I thought a set was more appropriate. The mfem native way of dealing with this is a vector of length maximum number of attributes that is zero if an attribute isn't active and 1 if it is. I thought this was much more user friendly.

**Comment by jamiebramwell** [comment]
LGTM

File: src/solvers/base_solver.cpp [hunk 2]
```diff
@@ -29,26 +29,25 @@ BaseSolver::BaseSolver(MPI_Comm comm, int n) : BaseSolver(comm)
...
-          bc->markers[i] = 0;
-          break;
-        }
+  bc->markers.SetSize(m_state.front()->mesh->bdr_attributes.Max());
+  bc->markers = 0;
+
+  for (int attr : ess_bdr) {
+    MFEM_ASSERT(attr <= bc->markers.Size(), "Attribute specified larger than what is found in the mesh.");
+    bc->markers[attr-1] = 1;
+    for (auto &existing_bc : m_ess_bdr) {
```
Comment (joshessman-llnl) [inline src/solvers/base_solver.cpp:44]: I'm not sure how big the "real" meshes get, but depending on the number of boundary conditions/attributes, the solver could also store a set of attributes already assigned to a boundary condition and avoid the linear search here.  

It might also worth be comparing `std::set` and `std::unordered_set` for the boundary attribute indications, but for small sets the difference is negligible.

**Review by joshessman-llnl — APPROVED** [review]

File: src/solvers/base_solver.cpp (hunk 2 above)
Comment (jamiebramwell) [reply src/solvers/base_solver.cpp:44]: These will typically be extremely small. However, this is really a stop gap measure until we have a real "entity set" type of boundary marker. The problem here is that the internal MFEM mesh format can only have one attribute marker or "color" per boundary face. In complex multiphysics simulations, you can easily have overlapping areas for different boundary conditions. Production codes use face or nodesets to get around this. Development of this capability is logged in this [issue](https://github.com/LLNL/serac/issues/12).

**Comment by jamiebramwell** [comment]
LGTM

**Comment by jamiebramwell** [comment]
This changes the user-facing interface for solvers to use std::sets instead of std::vectors of indicator values to denote the boundary attributes for applying boundary conditions
